server/parsec/components/memory/datamodel.py

```diff
@@ -79,6 +93,64 @@ class MemoryOrganization:
     # see https://github.com/Scille/parsec-cloud/pull/7324#discussion_r1616803899
     shamir_setup: dict[UserID, MemoryShamirSetup] = field(default_factory=dict)
 
+    # Stores topic name and discriminant (or `None`)
+    _topic_write_locked: set[TopicAndDiscriminant] = field(default_factory=set)
+    # Stores topic name and discriminant (or `None`) as key, and the number
+    # of concurrent read operations currently taking the lock as key
+    _topic_read_locked: dict[TopicAndDiscriminant, int] = field(
+        default_factory=lambda: defaultdict(lambda: 0)
+    )
+    _notify_me_on_topic_lock_release: Event | None = None
+
+    @asynccontextmanager
+    async def topics_lock(
+        self, read: Iterable[TopicAndDiscriminant] = (), write: Iterable[TopicAndDiscriminant] = ()
+    ):
+        while True:
+            # 1) Check the locks we want aren't already taken in write mode
+            if any(
+                topic_and_discriminant in self._topic_write_locked
+                for topic_and_discriminant in chain(read, write)
+            ):
+                if self._notify_me_on_topic_lock_release is None:
+                    self._notify_me_on_topic_lock_release = Event()
+                await self._notify_me_on_topic_lock_release.wait()
+                # The event we waited on doesn't specify which lock has been
+                # released, so we must retry from the beginning.
+                continue
+
+            # 2) Check the write locks we want aren't already taken in read mode
+            if any(
+                self._topic_read_locked.get(topic_and_discriminant, 0) != 0
+                for topic_and_discriminant in write
+            ):
+                if self._notify_me_on_topic_lock_release is None:
+                    self._notify_me_on_topic_lock_release = Event()
+                await self._notify_me_on_topic_lock_release.wait()
+                # The event we waited on doesn't specify which lock has been
+                # released, so we must retry from the beginning.
+                continue
+
+            # 3) All checks are good, we can now take the locks !
+
+            self._topic_write_locked.update(write)
+            for topic_and_discriminant in read:
+                self._topic_read_locked[topic_and_discriminant] += 1
+
+            yield
+
+            # 4) Release our locks and leave
+
+            self._topic_write_locked.difference_update(write)
+            for topic_and_discriminant in read:
+                self._topic_read_locked[topic_and_discriminant] -= 1
+
+            if self._notify_me_on_topic_lock_release is not None:
+                self._notify_me_on_topic_lock_release.set()
+                self._notify_me_on_topic_lock_release = Event()
```

Comment:
Use a `try-finally`, otherwise the lock won't be released when an exception is raised

Reply:
Good catch ! :1st_place_medal: 

fixed.